homeassistant/components/picnic/services.py

```diff
@@ -0,0 +1,131 @@
+"""Services for the Picnic integration."""
+from __future__ import annotations
+
+from typing import Any
+
+from python_picnic_api import PicnicAPI
+import voluptuous as vol
+
+from homeassistant.core import HomeAssistant, ServiceCall
+from homeassistant.helpers import device_registry
+import homeassistant.helpers.config_validation as cv
+
+from .const import (
+    ATTR_AMOUNT,
+    ATTR_DEVICE_ID,
+    ATTR_PRODUCT_ID,
+    ATTR_PRODUCT_IDENTIFIERS,
+    ATTR_PRODUCT_NAME,
+    CONF_API,
+    DOMAIN,
+    SERVICE_ADD_PRODUCT_TO_CART,
+)
+
+
+class PicnicServiceException(Exception):
+    """Exception for Picnic services."""
+
+
+async def async_register_services(hass: HomeAssistant) -> None:
+    """Register services for the Picnic integration, if not registered yet."""
+
+    # Register the add_product service
+    if not hass.services.has_service(DOMAIN, SERVICE_ADD_PRODUCT_TO_CART):
```

Comment:
Why is this needed? Because it runs once per config entry? I'm curious if this works to run from `async_setup` in `__init__.py` instead so it runs once on component init.

Reply:
I just looked at a few other components on how they implemented it and found this construct. (e.g. in the hue component) Is there a better way?

It's indeed to prevent registering services more than once.

Reply:
Thanks, hadn't seen it before.